Deque Analysis Color Contrast fixes

Dawnkai opened the pull request:
## References
* Fixes: #1149
* Fixes: #1159
* Fixes: #1161
* Fixes: #1163
* Fixes: #1202
* Fixes: #1194 (the final two color contrast issues)
* Fixes: #1178 (the final color contrast issue)

## Description
Fix incorrect color contrasts for various components in the UI. Some components need to have 3:1 contrast ratio, others 4.5:1.

## Instructions for Reviewers
Changes made by the PR:
1. Resolve https://github.com/DSpace/dspace-angular/issues/1149 by:
*  Making the `teal` color (used by hyperlinks) darker to match contrast requirement
2. Resolve https://github.com/DSpace/dspace-angular/issues/1159 by:
* Using `lighten` with `$gray-700` color (which has high contrast ratio) to change placeholder text color on inputs
* Darkening the color of `$cyan` color in Bootstrap to darken the color of days of the week in date popup
* Darkening the color of text in date popup
3. Resolve https://github.com/DSpace/dspace-angular/issues/1161 by:
* Darkening the color of the in-progress bar
* Using `darken` with Bootstrap's `$success` color to darken the color of success notification
4. Resolve https://github.com/DSpace/dspace-angular/issues/1163 by:
* Using `darken` with Bootstrap's `$green` color to darken the color of "Archived" badge on MyDspace page
* Using `darken` with Bootstrap's `$success` color to darken the color of success buttons
5. Resolve https://github.com/DSpace/dspace-angular/issues/1202 by:
* Increase color contrast of focus indicators on buttons by moving them outside of their border (using `outline` and `outline-offset` instead of box-shadow)
* Change colors of success, warning, danger and secondary Bootstrap button variants using `darken` combined with bootstrap's own color variables for both regular buttons and outline buttons
* Slightly (5-10%) increasing the opacity of the close icons on notification popups to increase contrast
* Using `darken` with Bootstrap's `$green` color to darken the color of the green bar in date field to meet contrast requirements
* Using Bootstrap's `$dark` color instead of `$gray` for borders of sliders on date field to meet contrast requirements
* Replacing box-shadow on `dynamic-ng-bootstrap-checkbox` with outline and `outline-offset` (same as for buttons in previous bullet points) to change contrast from checkbox's background and it's focus indicator to white background and focus indicator
* Using `darken` with `$danger` and `$success` colors for `.text-warning` and `.test-danger` class to meet contrast criteria for exclamation and check marks in submission form card headers

It might be worth considering changing bootstrap's default colors to avoid using `darken` and `lighten` or `adjust-hue`, however this PR changes colors only for elements specified in attached Issues in order to avoid potential conflicts in components that use Bootstrap's colors but were not mentioned in the Issues this PR closes.

While we we working on resolving the issue of insufficient color contrasts between inner background of a button and its focus indicator (i.e. see https://github.com/DSpace/dspace-angular/issues/1202 ID:468926) we came to the conclusion, based on UI/UX best practices, that it would be a better idea to move `:focus` box-shadows outside of the buttons by converting them into outlines. This leaves a blank space between the border of a button and its focus indicator. This way the only contrast reference is the white background (or red, when it comes to removed bitstreams on item edit) around the button. An exemplary explanation of this method is accessible here: https://craftcms.com/blog/designing-custom-focus-indicators.

Unfortunately, this means that in some places the focus indicator can leak onto other buttons. You can see the example of this below:
![MicrosoftTeams-image (3)](https://github.com/DSpace/dspace-angular/assets/30862280/e74efc83-d04b-4064-820d-c6c0ebd3bc5c)

However we consider this side effect to be acceptable, thus we propose this solution which will also prevent potential future contrast issues between the inside background colors of buttons and the `:focus` border colors.

**Include guidance for how to test or review your PR.** 
See the attached issues for details where to find html elements with incorrect contrast colors and see if their contrasts are correct now, i.e. by using a contrast checker available here: https://www.siegemedia.com/contrast-ratio.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [ESLint](https://eslint.org/) validation using `yarn lint`
- [x] My PR doesn't introduce circular dependencies (verified via `yarn check-circ-deps`)
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

tdonohue commented on src/styles/_custom_variables.scss:
```diff
@@ -24,14 +24,10 @@
   --ds-navbar-link-color-hover: #{darken($cyan, 15%)};
-  --ds-expandable-navbar-bg: var(--ds-navbar-bg);
-  --ds-expandable-navbar-link-color: var(--ds-navbar-link-color);
-  --ds-expandable-navbar-link-color-hover: var(--ds-navbar-link-color-hover);
```
This appears to be accidently removing CSS styles which are used in DSpace.  Why are all these styles being removed?

Dawnkai replied:
My apologies - I have restored the removed styles. I was working with another coworker on this PR and it seems like there was some issue with merging of our work. It has been resolved now.

tdonohue commented on src/styles/_custom_variables.scss:
```diff
@@ -24,14 +24,10 @@
   --ds-header-logo-height-xs: 50px;
   --ds-header-icon-color: #{$link-color};
   --ds-header-icon-color-hover: #{$link-hover-color};
-  --ds-navbar-bg: var(--ds-header-bg);
```
Additional style being accidently removed

tdonohue commented on src/styles/_global-styles.scss:
```diff
@@ -268,28 +269,125 @@ ul.dso-edit-menu-dropdown > li .nav-item.nav-link {
-  padding-right: 0.125rem !important;
-}
 
-.pb-0\.5 {
```
Three more styles here are accidently removed

tdonohue commented on src/styles/_global-styles.scss:
```diff
@@ -268,28 +269,125 @@ ul.dso-edit-menu-dropdown > li .nav-item.nav-link {
+		background-color: transparent;
+	}
+}
+ds-dso-edit-metadata-value {
```
Styles that are specific to an Angular component should be moved into the Component's stylesheet.  In this case, you seem to be changing styles for the "dso-edit-metadata-value" component, but you are making those styles difficult for others to find/change by putting them in the global stylesheet (`_global-styles.scss`)

In this situation, I think these styles should move to the stylesheet at `/src/app/dso-shared/dso-edit-metadata/dso-edit-metadata-value/dso-edit-metadata-value.component.scss`

Dawnkai replied:
I see. I will keep that in mind next time. I moved the component-specific styles to their respective component `.scss` files as you requested.

tdonohue commented on src/styles/_global-styles.scss:
```diff
@@ -268,28 +269,125 @@ ul.dso-edit-menu-dropdown > li .nav-item.nav-link {
+        opacity: 0.9;
+    }
+}
+ds-alert {
```
Same here, these styles need to be moved to our "ds-alert" component.

tdonohue commented on src/styles/_global-styles.scss:
```diff
@@ -268,28 +269,125 @@ ul.dso-edit-menu-dropdown > li .nav-item.nav-link {
+        opacity: 0.6;
+    }
+}
+ds-search-range-filter {
```
Same here, these styles need to be moved to our search-range-filter component.

tdonohue commented on src/styles/_custom_variables.scss:
```diff
@@ -61,7 +57,20 @@
+	--ds-button-success-background-color: #358726;
+	--ds-button-success-background-hover-color: #{darken(#358726, 10%)};
+	--ds-button-warning-background-color: #{darken($yellow, 20%)};
+	--ds-button-warning-background-hover-color: #{darken($yellow, 30%)};
```
I believe some of these button colors might be managed in `_bootstrap_variables.scss` as these are controllable in Bootstrap's own SCSS variables.  For instance, Bootstrap has a lot of SASS/SCSS variables for button colors, etc. https://github.com/twbs/bootstrap-sass/blob/master/assets/stylesheets/bootstrap/_variables.scss#L149-L182  

So, we might not need to override these manually.  It might be possible to just change the default Bootstrap variable(s) to make it a lot easier.  

If we use the default Bootstrap variables, you also might be able to remove some of the complex CSS that you added to `_global-styles.scss` for all these buttons. I haven't verified this myself though, but it does look like we are overriding a lot of Bootstrap defaults which might be easier to override if we just use the Bootstrap variables.

Dawnkai replied:
I think that's a great idea. However, when I tried to edit `_bootstrap_variables.scss` file according to the documentation of `bootstrap-sass` - i.e. by adding `$btn-danger-bg: #color` to it, it doesn't seem to cause any changes in the components. The same applies to theme colors in that file as well. It only affects `custom_variables.scss` file where these bootstrap variables are used.

I can control bootstrap colors in `src/themes/dspace/styles/_theme_sass_variable_overrides.scss`. Unfortunately, while I can change colors of the buttons by editing `$theme-colors` in that file, I am unable to change the color of, say, the danger button specifically using `$btn-danger-bg` variable (which is present in the file from the link you provided). I could technically make the changes using `$theme-colors`, however I am afraid it might affect other, non-button components.

Therefore I am not sure if this can be implemented - either because of me not having proper understanding of your comment or because there is an issue with bootstrap variable overriding.

artlowel replied:
@tdonohue the file you linked is from bootstrap-sass, that's no longer a separate project since a few versions, because bootstrap itself switched to sass. So @Dawnkai is right that there are no variables to set the color for each specific type of button. Nowadays everything is derived from the theme-map colors to make bootstrap easier to theme: you change the colors once there, and you affect everything.

However that does make something like this harder to do. I agree with @tdonohue that the way it happens in this PR is not ideal. Because all the colors are hardcoded here you can no longer change the success color in the theme map and have the buttons automatically change as well.

Ideally these hex values would still be derived from the theme colors using css or sass filters. That way you won't have to manually adjust these buttons as well every time you change the theme map colors. There will be no way to guarantee that the contrast will still be sufficient if someone changes those colors, but it will be a good starting point, and as long as the included dspace themes have sufficient contrast that'd be fine in my opinion.

Dawnkai replied:
@artlowel If I understand you correctly, while we can't really overwrite the bootstrap styles like @tdonohue proposed, we could replace hardcoded values in global styles and custom styles to values using the theme colors (i.e. instead of `#358726` for background color of a success button we would use something like `#{darken(map-get($theme-colors, success), 30%)}` or just `#{darken($success, 30%)}`.

If that's all we can do, I'll change the styles in this PR to use the theme colors. Is there anything you'd like me to change?

artlowel replied:
That's what I meant yes. 

I'd also move all the duplicated outline stuff

```scss
    box-shadow: none;
    outline: 2px solid ${somecolor};
    outline-offset: 2px;
```
… to one place in the css, so any changes to it only need to happen once, and then only override the colors in the success, warning, etc classes. Those colors would again ideally be derived from the theme colors.

tdonohue reviewed (changes requested):
@Dawnkai : First off, thanks for your contribution. I haven't tested this, but gave the changes a quick code review. 

It appears you are accidently removing some CSS settings in `_global-styles.scss`, and are also adding new CSS which is specific to components there.

Overall, the `_global-styles.scss` should only be used for styles that are necessary to change across the entire application.  Component specific styles should move into the corresponding `component.scss` file.  That makes it easier for institutions to know which styles each component is using.

More notes inline below.

tdonohue commented:
@artlowel : I'd appreciate it if you or someone on your team could help review the code here.  I'm not sure if there are easier ways to change some of these colors via Bootstrap SCSS variables instead of creating our own custom variables.  See for example [this discussion above](https://github.com/DSpace/dspace-angular/pull/2587#discussion_r1374686130).  I just want to better understand myself whether there are Bootstrap best practices here that we need to keep in mind.

Dawnkai commented:
@tdonohue : Hi, I updated the changes made by the PR by removing custom colors and replacing them with variants of Bootstrap's colors as requested. I also updated the PR's description to list all the changes made to fix contrasts requested in all the Issues this PR closes. I'd appreciate it if you could take a look and see if the PR is ready to deploy or not.

tdonohue reviewed (approved):
👍 Thank you @Dawnkai !  Apologies for taking so long to get back to this.  But, I finally tested this today, and I can verify that it fixes all the issues noted in the PR Description (I retested each scenario listed there using Deque's Axe DevTools plugin for web browsers).  All the color contrast issues appear to be resolved.

@artlowel : Did you want to take one last quick look at this? I believe all your feedback was addressed.  I'll leave it open for a day or two to give you a chance to look. After that I'll merge it both to 8.0 and 7.x

artlowel commented on src/app/shared/notifications/notification/notification.component.scss:
```diff
@@ -30,3 +31,9 @@
 .alert-warning .notification-progress-loader span {
   background: var(--ds-notification-bg-warning);
 }
+.alert-success{
+	color: darken($success, 22%);
+}
+.alert-danger {
+  color: darken($danger, 22%);
+}
```
If we use sass variables in component styles, they won't be affected by theme changes. That's why everything has been replaced with css variables: they do update on the fly. 

So I'd either turn these color into `ds-` variants and move them to custom_variables.scss, or, since these alert classes are global classes that aren't unique to the notification component anyway, it might make more sense to put these in global-styles.scss, where you _can_ use sass variables (you can anywhere inside the styles directory)

artlowel commented on src/app/shared/search/search-filters/search-filter/search-range-filter/search-range-filter.component.scss:
```diff
@@ -26,7 +26,7 @@
 
   .noUi-connect {
-      background: var(--ds-slider-color);
+      background: darken($green, 20%);
```
Same comment here regarding sass variables. Since it's a global class, I'd move it out of this component style

Dawnkai replied:
I restored the `--ds-slider-color` here as you requested. I also found out that I used `$dark` for handle slider, so I turned that into a css variable too (`--ds-slider-handle-color`).

tdonohue commented:
@artlowel : If you find a chance to verify your feedback has been addressed, I'd love to get this fixes into the codebase.  Thanks!

artlowel reviewed (approved):
Looks like all my feedback has been addressed.
Thanks @Dawnkai!

alanorth commented:
FYI I notice this PR has commits that add trailing whitespace. If possible it would be nice to trim those and rebase so we don't store them in the git history forever...

Dawnkai commented:
If I understood @alanorth properly, I removed the trailing whitespaces in the commit history. I also went ahead and fixed tabs to double space mismatch in global-styles.scss by replacing double whitespaces with a single tab (so that the style of indentation remains consistent).

alanorth commented on src/app/submission/form/footer/submission-form-footer.component.scss:
```diff
@@ -0,0 +1,15 @@
+	.btn:focus, .btn-outline-primary:not(:disabled):not(.disabled):active:focus, .btn-outline-primary:not(:disabled):not(.disabled).active:focus, .show > .btn-outline-primary.dropdown-toggle:focus, .custom-control-input:focus ~ .custom-control-label::before{
+		box-shadow: inset 0 3px 5px rgba(0, 0, 0, 0.25), 0 0 0 0.2rem rgba(27, 41, 55, 0.6);
+	}
+	
```
Trailing whitespace here.

alanorth commented on src/styles/_custom_variables.scss:
```diff
@@ -61,7 +61,20 @@
+  --ds-text-success-color: #74a030;
+  --ds-range-filter-border-color: #949494;
+  --ds-range-filter-connect-color: #63852e;
+	
```
Trailing whitespace here. This should simply be a newline, not a newline and a space/tab.

alanorth commented on src/styles/_custom_variables.scss:
```diff
@@ -61,7 +61,20 @@
+	--ds-button-success-background-hover-color: #{darken(#358726, 10%)};
+	--ds-button-warning-background-color: #{darken($yellow, 20%)};
+	--ds-button-warning-background-hover-color: #{darken($yellow, 30%)};
+	
```
Trailing whitespace here. This should simply be a newline, not a newline and a space/tab.

alanorth commented:
Thanks @Dawnkai. Maybe I wasn't clear enough. Several of your changes introduce "trailing whitespace", meaning a blank line with a space. Often times people fix this by adding a new commit to the pull request that fixes the whitespace, but then we end up with the whitespace and the fix in the git log! Better to just not commit trailing whitespace in the first place.

Regarding re-formatting the tabs to double spaces, I prefer git commits to be atomic and not touch things outside their scope, but it is not a policy in the DSpace code. Just my personal preference to keep things clean.

alanorth commented:
Thank you @Dawnkai. The changes are perfect now.
